promql/config/config.go

```diff
@@ -39,8 +39,10 @@ type QueryTweak struct {
 	NoBug                  bool                  `yaml:"no_bug,omitempty" json:"noBug,omitempty"`
 	TruncateTimestampsToMS int64                 `yaml:"truncate_timestamps_to_ms" json:"truncateTimestampsToMS,omitempty"`
 	AlignTimestampsToStep  bool                  `yaml:"align_timestamps_to_step" json:"alignTimestampsToStep,omitempty"`
+	AlignTimestampsToMS    int64                 `yaml:"align_timestamps_to_ms" json:"alignTimestampsToMS,omitempty"`
```

Comment:
Sorry, I actually had an error in my thinking when suggesting this as the option name and what it should do. In the end, an alignment to a specific value is logically not the same as forcing a certain number of digits to a specific value.

Thinking about this more, we could compose what you want via the existing `TruncateTimestampsToMS` tweak and then a new tweak type to add the desired value (999), for which we could name the query tweak `OffsetTimestampsByMS` (a value that can be set to a positive or negative number of ms that are then added to the timestamp).

Then you could use both query tweaks in the config file to achieve the desired effect:

```yaml
query_tweaks:
- note: "..."
  truncate_timestamps_to_ms: 1000
  offset_timestamps_by_ms: 999
```

Reply:
If we take this approach, the query tweak `OffsetTimestampsByMS `by itself will be mostly useless as we cannot predict the milliseconds in the query timestamps to add or truncate a specific value. And it would need to be always in conjunction with `TruncateTimestampsToMS` to use it. Also, to obtain the desired state of timestamp, vendors always need to use two query tweaks to solve one problem which is not right. 

As these query tweaks should go in conjunction, I have one more proposal to address this issue. We can use `TruncateTimestampsToMS` query tweak to have an optional offset parameter to add the milliseconds after truncating. There would be no use of adding one more query tweak `OffsetTimestampsByMS`

Changes may look like this
config.go

```
type QueryTweak struct {
	Note                   string                `yaml:"note" json:"note"`
	NoBug                  bool                  `yaml:"no_bug,omitempty" json:"noBug,omitempty"`
	TruncateTimestampsToMS int64                 `yaml:"truncate_timestamps_to_ms" json:"truncateTimestampsToMS,omitempty"`
	AlignTimestampsToStep  bool                  `yaml:"align_timestamps_to_step" json:"alignTimestampsToStep,omitempty"`
	DropResultLabels       []model.LabelName     `yaml:"drop_result_labels" json:"dropResultLabels,omitempty"`
	IgnoreFirstStep        bool                  `yaml:"ignore_first_step" json:"ignoreFirstStep,omitempty"`
	IgnoreCase             bool                  `yaml:"ignore_case" json:"ignoreCase,omitempty"`
	AdjustValueTolerance   *AdjustValueTolerance `yaml:"adjust_value_tolerance" json:"adjustValueTolerance,omitempty"`
}
```
```

type TruncateTimestampsToMS struct {
	TruncateMS       *float64 `yaml:"truncate_ms" json:"truncateMS,omitempty"`
	OptionalOffsetMS *float64 `yaml:"optional_offset_ms" json:"optionalOffsetMS,omitempty"`
}
```
expand.go

```
func applyQueryTweaks(tc *comparer.TestCase, tweaks []*config.QueryTweak) *comparer.TestCase {
	resTC := *tc
	for _, t := range tweaks {
		if t.TruncateTimestampsToMS != nil {
			truncateDuration := time.Duration(0)
			offsetDuration := time.Duration(0)
			if t.TruncateTimestampsToMS.TruncateMS != nil {
				truncateDuration = time.Duration(*t.TruncateTimestampsToMS.TruncateMS) * time.Millisecond
			}
			if t.TruncateTimestampsToMS.OptionalOffsetMS != nil {
				offsetDuration = time.Duration(*t.TruncateTimestampsToMS.OptionalOffsetMS) * time.Millisecond
			}
			resTC.Start = resTC.Start.Truncate(truncateDuration).Add(offsetDuration)
			resTC.End = resTC.End.Truncate(truncateDuration).Add(offsetDuration)
		}
		if t.AlignTimestampsToStep {
			resTC.Start = resTC.Start.Truncate(resTC.Resolution)
			resTC.End = resTC.End.Truncate(resTC.Resolution)
		}
	}
	return &resTC
}
```

By this, vendors have the flexibility to provide or not provide any of the values as required for `TruncateTimestampsToMS` tweak


And config file can go as:
```
  - note: 'Truncate query timestamps to MS and add optional offset MS'
    truncate_timestamps_to_ms:
      truncate_ms: 1000
      optional_offset_ms: 999
```

Reply:
Yeah, that would be fine with me as well, it just has the downside that more people will need to change their configs to adjust to the breaking change in configuration.

Due to that reason I think I still *slightly* prefer the compositional variant, as it wouldn't actually show up as two separate query tweaks for a vendor in the result: Both options can be set in the same single query tweak entry with the same note:

```
query_tweaks:
- note: "A single note describing what we are tweaking"
  truncate_timestamps_to_ms: 1000
  offset_timestamps_by_ms: 999
```

Reply:
If we can provide two options and consider it as a single query tweak under same note, vendors can document it as a single cross cutting issue. 

Made all the required changes as per comments. Can add test.yml file when ready for official conformance testing.